[ios] Add `Open In App ...` button to the PlacePage

kirylkaveryn opened the pull request:
Closes https://github.com/organicmaps/organicmaps/issues/7174

Design https://www.figma.com/design/tbIeaE9ixXpDW875A3gpd6/OM-ShareToApp?node-id=0-1&t=AMq3nRXB06N4NYCC-1

This PR implements an additional button to open the poi in the app selected from the list.

### Key features:
- the last used app is displayed in the PP with the name and icon right below the Coordinates cell
- if the user taps on it - the target app will be opened with the provided poi coordinates
- by tapping on the accessory button on the right the sheet with the available apps list will be displayed
- if user select an app from the list, the app will be opened and will be saved as the `last used` and displayed in PP

### List of available apps
- [x] Google Maps
- [x] Apple Maps
- [x] OsmAnd - should we actually add this one?
- [x] Yandex Maps
- [x] 2GIS
- [x] CityMapper
- [x] Moovit
- [x] Uber
- [x] Waze

### Tested on:
- [x] iOS 17.5 iPhone (device)
- [x] iOS 15.5 iPhone (simulator)
- [x] iOS 17.5 iPad (simulator)
- [x] iOS 12.5 iPhone 6 (device)
- [x] iPadDesignedForMac

___
### Problems:
Unless the design was discussed and OK in the real-world usage I found that (especially on I pad when thw whole PP is always visible) that always displays the previously selected app on the PP **looks like an advertisement** and distracts a lot when you select different pois.
Its like if you 5 times select pois you 5 times will see the "Apple Maps" or "Super Maps" - its really annoying...

IMO it would be better to use a some _standart_ symbol as an icon (as @oleg-rswll already suggested) and not write the app name - just leave "Open In App...". It would be better to open the sheet by tapping on the entire button every time. I think that it is not so often used operation that we should always show the last used app in the PP. Also it will prevent accidental taps.

@biodranik @oleg-rswll @rtsisyk please share your thoughts!

(screenshots are not super fresh and "Open In" is used in translations instead of "Open With")

<img width="350" alt="image" src="https://github.com/organicmaps/organicmaps/assets/79797627/f0e2da36-dbe8-4d93-bffa-735b71e67b0f">

<img width="350" alt="image" src="https://github.com/organicmaps/organicmaps/assets/79797627/f47f3302-b4f2-4771-a0c8-57da569f3dfc">

https://github.com/organicmaps/organicmaps/assets/79797627/70f0b654-089f-4f73-92e0-a67c3f7eaaae

<img width="600" alt="image" src="https://github.com/organicmaps/organicmaps/assets/79797627/8c17728c-a94e-46bc-8ff7-ae80c1fa2b68">

euf commented:
Absolutely need **Waze**, since I imagine the main use case to be using live traffic navigation. They have documentation on deep links: https://developers.google.com/waze/deeplinks/

Also I agree with @oleg-rswll that a generic "Open in…" icon would be better than app icons. 

I designed one, please take a look: 

![0-open-in](https://github.com/organicmaps/organicmaps/assets/11754264/ca8eaefe-cc54-48e8-9d64-ba5645be9a8f)

Arthur-GYT commented on data/strings/strings.txt:
```diff
@@ -30687,3 +30687,49 @@
+    eu = Ireki %@-n
+    fa = باز کردن در %@
+    fi = Avaa %@
+    fr = Ouvrir en %@
```
If the translation "Open In ..." remains, in French it would be more "Ouvrir dans ..." but then if it's the translation "Open with..." (which I think would be better), it would be more like "Ouvrir avec...".

Arthur-GYT replied:
@organicmaps/translations-fr  Do you have a different opinion?

kirylkaveryn replied:
@Arthur-GYT Thanks! But this in not the final version of the title. Ill mention translators when change the PR from draft to ready for review

patepelo commented:
Would "Open in another app" fit better?

kirylkaveryn commented:
The latest implementation:

https://github.com/organicmaps/organicmaps/assets/79797627/0abc6b65-15a6-49f8-9639-4a0712d9c1f4

<img width="350" alt="image" src="https://github.com/organicmaps/organicmaps/assets/79797627/1c2edbda-8b87-4473-9496-ddb71a28fca6">
<img width="350" alt="image" src="https://github.com/organicmaps/organicmaps/assets/79797627/00fe0a9a-2f36-48ab-ae97-b2eac5f141f2">

oleg-rswll commented:
The icon looks bigger than other icons in the Place Page, probably good to make it a bit smaller to be sized like the others.

kirylkaveryn commented:
@euf can you please take a look? Icon you provide is great but I agree with @oleg-rswll - maybe a little too big in comparison with the others... I've constrained them all to 28x28pt.

biodranik commented:
1. Why 2gis? 
2. Why uber without an affiliate link?

kirylkaveryn commented:
Let's finish with the applications list.

The latest version looks like:
<img width="300" alt="image" src="https://github.com/organicmaps/organicmaps/assets/79797627/c5e29304-f78f-4fb0-8caa-09ccc7712ea8">
<img width="300" alt="image" src="https://github.com/organicmaps/organicmaps/assets/79797627/c334a29c-da93-444f-8aab-cd357bfd1322">

dvdmrtnz commented:
Why is wheelchair icon tiny?

> Let's finish with the applications list.
> 
> The latest version looks like: <img alt="image" width="300" src="https://private-user-images.githubusercontent.com/79797627/345098848-c5e29304-f78f-4fb0-8caa-09ccc7712ea8.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk5OTYwNTIsIm5iZiI6MTcxOTk5NTc1MiwicGF0aCI6Ii83OTc5NzYyNy8zNDUwOTg4NDgtYzVlMjkzMDQtZjc4Zi00ZmIwLThjYWEtMDljY2M3NzEyZWE4LlBORz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzAzVDA4MzU1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM4YzEzOTFlZjZmMjc5ODQ2NzU4NDAxZmJlYjgxNmE5YmQxZDg3YjlmOTA3MDg2Njc5ZTY3Nzg4NDgyZmE0MzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.69aMp4gGLgpK6t40NlfarGddJLPujhgnBbP77SgVdeo"> <img alt="image" width="300" src="https://private-user-images.githubusercontent.com/79797627/345098878-c334a29c-da93-444f-8aab-cd357bfd1322.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk5OTYwNTIsIm5iZiI6MTcxOTk5NTc1MiwicGF0aCI6Ii83OTc5NzYyNy8zNDUwOTg4NzgtYzMzNGEyOWMtZGE5My00NDRmLThhYWItY2QzNTdiZmQxMzIyLlBORz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzAzVDA4MzU1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQwOTc2ZGI3ODdiMTJkZGVkMzU2ZWRmZGMxZWMxY2M3MDgwNGFkODBjZWZjZTRiMDI1ZTYxZmFiZWFlMmNiMDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.VFkccwDJdpQopp2dcgQqa_41OhqirePlgaOjvKbpEMY">

NeatNit commented:
`he` looks good. This would be cool to have on Android as well.

Jean-BaptisteC commented:
@rtsisyk please check, maybe we can doing this same implementation on Android instead a button under the place page

soshial commented on iphone/Maps/UI/PlacePage/PlacePage.storyboard:
I am surprised that this is not a part of "Share" button. Those 2 flows can be easily merged into a single one. And we wouldn't clutter the place page view.

kirylkaveryn replied:
There are multiple solutions and they have the pros and cons that was discussed in the related issue and provided design.

Please share an example of how to merge these two flows.

soshial replied:
Sorry, I didn't know about this. I like variant 2.1 and 2.2.
Those save a lot of space.

biodranik replied:
@soshial opening specific osm object is possible only after querying OSM data and matching the object. No OSM ids are stored in mwm files.

jimcarst reviewed (commented):
nl is ok

kirylkaveryn commented:
> Why is wheelchair icon tiny?
> 
> >

Thanks for your attention! This is because the icon wasn't properly designed and have too big insets (usually we use the 28x28p frame and 24x24 image size):
<img width="200" alt="image" src="https://github.com/organicmaps/organicmaps/assets/79797627/2094ac2a-7a98-4a4e-b257-be167c1f45a7">

In this PR I changed the images content mode from the `center` to `aspectFit`:

Fixed:
<img width="350" alt="image" src="https://github.com/organicmaps/organicmaps/assets/79797627/8b14dced-28d3-4ce6-b6d8-60dc5d7abf74">

dvdmrtnz commented:
Place page attributes are no longer multiline after these changes. Multiline attributes were added in https://github.com/organicmaps/organicmaps/pull/7271

<img src="https://github.com/organicmaps/organicmaps/assets/47610359/186ab4eb-3235-486c-a206-1d548846c41d" width="320"/>

pastk commented:
How is it different from the existing "Share" button?

Does it make sense / is feasible to combine them?

re supported app list - does it have to be hardcoded?
(and I also don't get what is the special concern with Osmand?)

kirylkaveryn commented:
> How is it different from the existing "Share" button?
> 

In iOS there is no possibility to modify the default share screen because it is fully composed by the system.
Open in app != share the link/text so the opening app is different feature.

> Does it make sense / is feasible to combine them?
> 

We discussed with @biodranik and @rtsisyk that in ios it is not the good solution.

> re supported app list - does it have to be hardcoded?

Yes. 

> (and I also don't get what is the special concern with Osmand?)

Agree with you. @biodranik WDYT?

euf commented:
Please use these visually balanced and improved icons for the "Open In" and for  wheelchair access:

![open-in-1](https://github.com/organicmaps/organicmaps/assets/11754264/7e43e6d6-18b5-4344-89ff-5df3d5f4a6c6)

![wheelchair-1](https://github.com/organicmaps/organicmaps/assets/11754264/5f7993c7-27dd-4883-bc32-67a1c7516523)

kirylkaveryn commented:
> Place page attributes are no longer multiline after these changes. Multiline attributes were added in #7271

Sorry for that. Fixed:

<img width="350" alt="image" src="https://github.com/organicmaps/organicmaps/assets/79797627/ff83a312-86d9-4032-b15d-50888e1c8367">


> Please use these visually balanced and improved icons for the "Open In" and for wheelchair access:
Lovely icons! Thank you very much! 

<img width="350" alt="image" src="https://github.com/organicmaps/organicmaps/assets/79797627/16df621e-3071-4411-914d-7ae33d1020d9">

biodranik commented on iphone/Maps/UI/PlacePage/Components/OpenInAppActionSheet/OpenInApplication.swift:
```diff
@@ -0,0 +1,114 @@
+    case .appleMaps:
+      return "Apple Maps"
+    case .osmAnd:
+      return "OsmAnd Maps"
```
OsmAnd

kirylkaveryn replied:
But this app is actually called `OsmAnd Maps` in iOS

![image](https://github.com/organicmaps/organicmaps/assets/79797627/1f1bf35c-b89c-4730-94e8-9d1610fcf5c3)

biodranik commented on iphone/CoreApi/CoreApi/Common/MWMGeoUtil.mm:
```diff
@@ -58,4 +58,9 @@ + (float)angleAtPoint:(CLLocationCoordinate2D)p1 toPoint:(CLLocationCoordinate2D
 }
 
++ (NSString *)formattedOsmLinkForCoordinate:(CLLocationCoordinate2D)coordinate {
+  auto const link = measurement_utils::FormatOsmLink(coordinate.latitude, coordinate.longitude, 14);
```
The current zoom level from the map should be used instead of hard-coded 14.

biodranik commented on iphone/Maps/UI/PlacePage/Components/OpenInAppActionSheet/OpenInApplication.swift:
```diff
@@ -0,0 +1,114 @@
+    case .appleMaps:
+      return "\(scheme)?q=\(latitude),\(longitude)"
+    case .osmAnd:
+      return "\(scheme)?lat=\(latitude)&lon=\(longitude)&z=15"
```
The current zoom level should be used instead of a hard-coded one.

biodranik commented on iphone/Maps/UI/PlacePage/Components/OpenInAppActionSheet/OpenInApplication.swift:
```diff
@@ -0,0 +1,114 @@
+    case .osmAnd:
+      return "\(scheme)?lat=\(latitude)&lon=\(longitude)&z=15"
+    case .yandexMaps:
+      return "\(scheme)maps.yandex.ru/?pt=\(longitude),\(latitude)&z=16"
```
Ditto

biodranik commented on iphone/Maps/UI/PlacePage/Components/OpenInAppActionSheet/OpenInApplication.swift:
```diff
@@ -0,0 +1,114 @@
+    case .dGis:
+      return "\(scheme)2gis.ru/geo/\(longitude),\(latitude)"
+    case .cityMapper:
+      return "\(scheme)directions?endcoord=\(latitude),\(longitude)&endname="
```
endname can be initialized from the place page title.

kirylkaveryn replied:
for the different apps, it works differently so the code is updated to add the name where it's possible (for google maps it doesn't).

biodranik commented on iphone/Maps/UI/PlacePage/Components/OpenInAppActionSheet/UIAlertController+openInAppActionSheet.swift:
```diff
@@ -0,0 +1,26 @@
+                                      didSelectApp: @escaping OpenInApplicationCompletionHandler) -> UIAlertController {
+    let alertController = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
+
+
```
extra new line

biodranik reviewed (commented):
Thanks! @fabwu @dvdmrtnz can you please help with review and testing?

fabwu commented on iphone/CoreApi/CoreApi/Common/MWMGeoUtil.h:
```diff
@@ -18,6 +18,7 @@ NS_SWIFT_NAME(GeoUtil)
 @interface MWMGeoUtil : NSObject
 
 + (float)angleAtPoint:(CLLocationCoordinate2D)p1 toPoint:(CLLocationCoordinate2D)p2;
++ (NSString *)formattedOsmLinkForCoordinate:(CLLocationCoordinate2D)coordinate zoomLevel:(int)zoomLevel;
```
Where do you use this function? I can't find any reference to it...

fabwu replied:
Uff I didn't know that 😅 Thanks!

fabwu reviewed (commented):
Tested on my iPhone 17.5 (Openstreet Maps, Google Maps and Uber) and it works great :)

kirylkaveryn commented on iphone/Maps/UI/PlacePage/Components/OpenInAppActionSheet/OpenInApplication.swift:
```diff
@@ -0,0 +1,123 @@
+  case waze
+}
+
+extension OpenInApplication {
```
The key point of my solution is:
1. the data associated with the target application should be defined only the one time because its impossible situation that there can be >1 same application. This is why enum is better approach -  it isolates creational and usage process
2. every app always have the same set of options

You approach with the structs pushses to create a separate type for the every new application.

fabwu replied:
My main concern was that you could miss adding a switch case statement when you introduce a new case for the enum. I've just realized that this produces a compiler error in swift, so your solution is fine.

I'm used to enums in Java but they are apparently different to enums in swift.

kirylkaveryn commented on data/strings/strings.txt:
```diff
@@ -30618,3 +30618,49 @@
     zh-Hant = 錯誤：iCloud 不可用
+
+  [open_in_app]
+    comment = Title for the "Open In Another App" buton on the PlacePage.
```
thanks! fixed

kirylkaveryn commented on iphone/Maps/UI/PlacePage/Components/OpenInAppActionSheet/OpenInApplication.swift:
```diff
@@ -0,0 +1,123 @@
+    }
+  }
+
+  func linkWith(coordinates: CLLocationCoordinate2D, zoomLevel: Int = Int(FrameworkHelper.currentZoomLevel()), destinationName: String? = nil) -> String {
```
I don't understand why you are confused by the default value for the argument...
This `FrameworkHelper.currentZoomLevel()` will be used only when there are no special value are provided by the caller.
There is no reason to create the default value for the zoom level every time directly.
This solution composes the benefits of the two solution you described:
1. the argument isn't visible while calling the method if you have any special behaviour
2. the dependency is moved (semantically) creation of the dependency out of the scope

biodranik replied:
Does it make sense to retrieve the zoom level only when it is necessary?

kirylkaveryn replied:
zoom levl is retrived only when the link is built ==  when it is needed
It is not stored property.

fabwu replied:
I just wouldn't add parameters to a function which aren't used by a caller at the moment . For me if a parameter isn't used it turns into a local variable, but maybe I'm a bit to strict about that :)

kirylkaveryn commented:
@biodranik maybe we can add this one into the next alpha?

kirylkaveryn commented on drape_frontend/drape_engine.cpp:
```diff
@@ -166,6 +166,11 @@ void DrapeEngine::SetVisibleViewport(m2::RectD const & rect) const
 
+int DrapeEngine::GetCurrentZoomLevel() const
+{
+  return m_frontend->GetCurrentZoom();
```
The zoom level isn't added to the shared link.
https://github.com/organicmaps/organicmaps/blob/0363d87cc3b3db47abeb51f76774e3d1483bb508/iphone/Maps/Classes/Share/MWMShareActivityItem.mm#L54-L82
The best zoom is calculated during the deeplink handler for the **incoming** link:

https://github.com/organicmaps/organicmaps/blob/0363d87cc3b3db47abeb51f76774e3d1483bb508/map/mwm_url.cpp#L420C3-L436

There were no public methods to get the current zoom level. It was creates as you request.

biodranik replied:
What about `Framework::GetDrawScale()`? It is used for short url generation:

```
std::string const & s = f.CodeGe0url(ll.m_lat, ll.m_lon, f.GetDrawScale(), title(self.data).UTF8String);
```

kirylkaveryn replied:
Thanks! Quite not obvious naming in terms of `zoom`...

kirylkaveryn replied:
@biodranik fixed!

kirylkaveryn commented on drape_frontend/frontend_renderer.hpp:
```diff
@@ -351,12 +352,6 @@ class FrontendRenderer : public BaseRenderer,
     return m_currentZoomLevel >= 0;
   }
 
-  int GetCurrentZoom() const
```
Because it was private method. But it should be public to retrieve the current zoom.

kirylkaveryn commented on iphone/Maps/UI/PlacePage/Components/OpenInAppActionSheet/OpenInApplication.swift:
```diff
@@ -0,0 +1,123 @@
+    }
+  }
+
+  func linkWith(coordinates: CLLocationCoordinate2D, zoomLevel: Int = Int(FrameworkHelper.currentZoomLevel()), destinationName: String? = nil) -> String {
```
It will increase the overall code amount and implementation time but may save some time for the android because:
- it needs to implement all the logic into the c++ 
- objc-c wrapper for c++ (it will be great to use swift/c++ interop one day)
- jni wrappers for c++

biodranik replied:
Ok, let's raise this question later when implementing a similar functionality for Android. It would be great to create a new issue (or modify the existing one) to leave it open for the missing Android implementation.

biodranik commented on iphone/Maps/OMaps.plist:
```diff
@@ -92,6 +92,14 @@
 	<key>LSApplicationQueriesSchemes</key>
 	<array>
 		<string>mailto</string>
```
CanOpenUrl is also called for `ms-outlook://` and `googlegmail://` schemes. Should they be added into this list too?

biodranik reviewed (approved):
Thanks!
